alica_engine/src/engine/scheduler/Scheduler.cpp

```diff
@@ -0,0 +1,107 @@
+#include "engine/scheduler/Scheduler.h"
+#include "engine/AlicaEngine.h"
+
+#include <algorithm>
+
+namespace alica
+{
+
+namespace scheduler
+{
+
+Scheduler::Scheduler(int numberOfThreads, const alica::AlicaEngine* ae) : _ae(ae), _running(true)
+{
+    for (int i = 0; i < numberOfThreads; i++) {
+        _workers.push_back(new std::thread(&Scheduler::workerFunction, this));
+    }
+}
+
+Scheduler::~Scheduler()
+{
+    _running = false;
+    _jobQueue.clear();
+
+    for (auto worker : _workers) {
+        _workerCV.notify_all();
+        worker->join();
+        delete worker;
+    }
+}
+
+void Scheduler::schedule(std::shared_ptr<Job> job)
+{
+    if (job->scheduledTime.inNanoseconds() == 0) {
+        job->scheduledTime = _ae->getAlicaClock().now();
+    }
+
+    if (job->inExecution || _jobQueue.isScheduled(job)) {
+        return;
+    }
+
+    // schedule missing prerequisites
+    for (auto prerequisite : job->prerequisites) {
+        if (auto prerequisiteSharedPtr = prerequisite.lock())  {
+            schedule(prerequisiteSharedPtr);
+        }
+    }
+
+    _jobQueue.insert(job);
+    _workerCV.notify_one();
+}
+
+void Scheduler::workerFunction()
+{
+    while(_running.load()) {
+        std::shared_ptr<Job> job;
+        bool executeJob = true;
+
+        {
+            std::unique_lock<std::mutex>lock(_workerMtx);
+            // wait when queue is empty. Do no wait when queue has jobs or the schedulers destructor is called.
+            _workerCV.wait(lock, [this]{return !_jobQueue.isEmpty() || !_running.load();});
+
+            if (_jobQueue.isEmpty() || !_running.load()) {
+                continue;
+            }
+
+            job = _jobQueue.popNext();
+
+            if (job->cancelled) {
+                /*
+                 * If job is cancelled:
+                 *  - Dont execute job.
+                 *  - Dont schedule job.
+                 */
+                continue;
+            }
+
+            job->inExecution = true;
+
+            for (std::weak_ptr<Job> prerequisite : job->prerequisites) {
+                if (prerequisite.lock()) {
+                    // Prerequisite jobs not finished, do not execute job
+                    executeJob = false;
+                    break;
+                }
+            }
+        }
+
+        if (executeJob) {
+            try {
+                job->cb();
+                if (job->isRepeated) {
+                    job->inExecution = false;
+                    schedule(job);
+                }
+            } catch (std::bad_function_call& e) {
+                std::cerr << "ERROR: Bad function call\n";
+            }
+        } else {
+            job->inExecution = false;
+            schedule(job);
```

Comment:
popping a task only to potentially reschedule is is rather expensive. Also you are only considering the first task in the queue, the second may be executable.

Try the following:
- find the first free queue element from within the lambda you pass to `wait(...)`, it is executed under a mutex
- remove that from the queue (move onto stack)
- drop the mutex lock (as you do now)
- execute

Reply:
There is a fix for just rescheduling it now, since this would have been essentially busy waiting in some cases. 

However, looking at more than the first task further improves this. @Karasuma1412 please go for it.

Reply:
Why would there be busy waiting?

Reply:
The original code would have taken out the first task and put it back into the queue over and over again, although it wasn't ready for execution. -> That is what I denoted busy waiting.